### ISSUE TITLE
Optional parameter to do hard power_cycle

### ISSUE THREAD

**Issue by henn** [CCI-MOC/hil#631]
Currently, the [reset API](https://github.com/CCI-MOC/hil/blob/master/docs/rest_api.md#node_power_cycle) does a "soft reset", based on [this code](https://github.com/CCI-MOC/hil/blob/master/haas/ext/obm/ipmi.py#L68):

``` python
    @no_dry_run
    def power_cycle(self):
        self._ipmitool(['chassis', 'bootdev', 'pxe'])
        if self._ipmitool(['chassis', 'power', 'cycle']) == 0:
            return
        if self._ipmitool(['chassis', 'power', 'on']) == 0:
            # power cycle will fail if the machine is not running.
            # To avoid such a situation, just turn it on anyways.
            # Doing this saves power by turning things off without
            # Without breaking the HaaS.
            return
        # If it is still does not work, then it is a real error:
        raise OBMError('Could not power cycle node %s' % self.node.label)
```

We should introduce an optional parameter (like `force`) that does an `ipmitool chassis power reset`. Here's what the ipmitool man page says about the different states:

```
Performs a chassis control  command  to  view  and change the power state.

status

       Show current chassis power status.

on

       Power up chassis.

off

       Power  down  chassis  into  soft off (S4/S5 state).  WARNING: This command does not
       initiate a clean shutdown of the operating system prior to powering down  the  sys‐
       tem.

cycle

       Provides  a  power  off  interval  of at least 1 second.  No action should occur if
       chassis power is in S4/S5 state, but it is recommended to check power  state  first
       and  only issue a power cycle command if the  system  power is on or in lower sleep
       state than S4/S5.

reset

       This command will perform a hard reset.

diag

       Pulse a diagnostic interrupt (NMI) directly to the processor(s).

soft

       Initiate a soft-shutdown of OS via ACPI.  This can be done in  a  number  of  ways,
       commonly by simulating an overtemperture or by simulating a power button press.  It
       is necessary for there to be Operating System support for ACPI  and  some  sort  of
       daemon watching for events for this soft power to work.
```

**Comment by henn** [comment]
@apoorvemohan and @ravisantoshgudimetla are interested in this so they can reset machines faster than the current implementation.

**Comment by matsuuran** [comment]
@henn do you think I should have two cli commands `node_power_cycle` and `node_power_reset` or should I just add an option to pass from the cli -> api -> ipmi?  The reason I ask is because the nomenclature might be a little confusing if there is only one command because the man page specifically calls out cycle and reset.

Should be simple either way, just wanted to get your input.

**Comment by matsuuran** [comment]
@henn any opinion?

**Comment by henn** [comment]
Thanks for looking at this, @matsuuran!

Ideally, the CLI would take an optional parameter to node_power_cycle (maybe `--force`, `--fast` and `-f` for short), since the CLI doesn't need to match the API exactly. To do that, though, would require implementing better CLI handling (see #431).

**Comment by naved001** [comment]
Quoting @zenhack from #372 because he brought up an important point there.

> When doing integration tests for the HaaS release, @lokI8 and I noticed that the sun blades were shutting down properly when we invoked the power cycle api, if they were booted into CentOS. We should make sure it isn't possible for a malicious user to simply "ignore" the shtudown request, thereby hijacking a node that they are supposed to have released.

> Apparently the hard reset command is chassis power reset whereas we're calling chassis power cycle.

**Comment by henn** [comment]
Per discussion with @naved001 and @Izhmash, we probably want to keep the default as "soft reset". We can change the default later if we decide we want that.

**Comment by zenhack** [comment]
Can you summarize the discussion? I don't have a strong opinion, but I'd like to have a paper trail for technical decisions like this.

**Comment by naved001** [comment]
As far as I remember the discussion was just that let's do soft reset by default.

**Comment by zenhack** [comment]
Ok, so basically you guys just picked one, without terribly specific reasoning? (I'm fine with this, but just want to make sure the reasoning is recorded)

**Comment by Izhmash** [comment]
The choice wasn't super important, but we chose soft as a default because hard reset can potentially cause damage.  Another reason is so users can keep using the command as they used to.  When omitting the `force` parameter, it will function will operate just the same as before.

**Comment by zenhack** [comment]
Cool, works for me.

**Comment by Izhmash** [comment]
Fixed by #835
